test/kubernetes/integration_test.go

```diff
@@ -132,414 +153,481 @@ func TestPod_Single_Volume(t *testing.T) {
 			StorageClassName: strPtr(testStorageClass),
 		},
 	}
+}
 
-	t.Log("Creating pvc")
-	_, err = client.CoreV1().PersistentVolumeClaims(namespace).Create(pvc)
-	if err != nil {
-		t.Fatal(err)
-	}
+// testAppName sanitizes the appName based off of the given test name
+func testAppName(t *testing.T) string {
+	return strings.ToLower(strings.ReplaceAll(t.Name(), "_", "-"))
+}
 
-	t.Logf("Waiting pod %q to be running ...\n", pod.Name)
-	if _, err := waitForPod(client, pod.Name); err != nil {
-		t.Error(err)
+// testClaimName sanitizes the claim name based off the app name
+func testClaimName(t *testing.T) string {
+	return fmt.Sprintf("pvc-%s", testAppName(t))
+}
+
+func TestPod_Single_Volume(t *testing.T) {
+	t.Parallel()
+	appName := testAppName(t)
+	volumeName := "my-do-volume"
+	claimName := testClaimName(t)
+
+	tt := []struct {
+		pod func() *v1.Pod
+		pvc func() *v1.PersistentVolumeClaim
```

Comment:
I remember we discussed that the integration tests will only be useful for testing Kubernetes < 1.14 going forward because newer releases support running the upstream end-to-end tests (that ship with an entire suite of access mode-related tests). What we likely want at some point is introduce a flag to run all integration tests for older releases (1.13 while still supported by us, potentially even older ones) and none for newer releases.

I understand the function returning a PVC is to support testing for filesystem and block mode, respectively, here and for most (all?) test we have build so far. Unfortunately, this would make it difficult to run the tests on 1.13 or below because [block mode turned beta in Kubernetes 1.14 only](https://kubernetes-csi.github.io/docs/raw-block.html); meaning that the block mode-related tests would all fail on older releases.

Thinking about how we can resolve this, I can see two possibilities:

1. Test the access mode in a single, dedicated test only and revert the pre-existing tests to their prior state. That way, we can easily turn tests on and off as needed per tested Kubernetes release.
1. Do not test the access mode in our integration tests at all but rely on the upstream tests only. Merging the feature without any tests is not ideal, however, so for this scenario we'd probably want to wait for #248 first so that we can enable the block testing capability in the upstream tests.

The first option allows us to move forward without depending on #248, at the price of a bit more tech debt (as we eventually will be able to remove our access mode integration tests entirely in favor of the upstream ones; at least, I assume so). #248 seems close to completion, but you never know, so maybe option 1. seems slightly preferable.

Let me know what you think.

Reply:
Good catch on the version support :+1: 

I think #248 is definitely the direction we will move towards but I think `1` or a variation of it would be better in the short term to reduce the risk of this PR getting blocked. 

I'm fine either creating a separate test as you suggested in number `1` or simply calculating the `kubernetes_version` of the cluster in test during `setup` and then having a minimum version gate for the table driven test configuration. I think either way it would be smart to create a version gate for the block tests.

I'm personally leaning towards adding the gate into the updated table driven tests for the enhanced test cleanup functionality and naming. I think if they were staying long term it would be the right decision, but since these tests also have a relatively short EOL I would be fine with reverting them to prior state as well.

Reply:
Gating on top of your improvements sounds fine for me as well. 👍